sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala

```diff
@@ -142,3 +196,414 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
+  extends GenericArrayData {
+  override def array(): Array[Any] = primitiveArray.toArray
+
+  override def copy(): ArrayData = new GenericShortArrayData(primitiveArray)
+
+  override def numElements(): Int = primitiveArray.length
+
+  override def isNullAt(ordinal: Int): Boolean = false
+  override def getShort(ordinal: Int): Short = primitiveArray(ordinal)
+  override def toShortArray(): Array[Short] = {
+    val array = new Array[Short](numElements)
+    System.arraycopy(primitiveArray, 0, array, 0, numElements)
+    array
+  }
+  override def toString(): String = primitiveArray.mkString("[", ",", "]")
+
+  override def equals(o: Any): Boolean = {
+    if (!o.isInstanceOf[GenericShortArrayData]) {
+      return false
+    }
+
+    val other = o.asInstanceOf[GenericShortArrayData]
+    if (other eq null) {
+      return false
+    }
+
+    val len = numElements()
+    if (len != other.numElements()) {
+      return false
+    }
+
+    var i = 0
+    while (i < len) {
+      val o1 = primitiveArray(i)
+      val o2 = other.primitiveArray(i)
+      if (o1 != o2) {
+        return false
+      }
+      i += 1
+    }
+    true
+  }
+
+  override def hashCode: Int = {
+    var result: Int = 37
+    var i = 0
+    val len = numElements()
+    while (i < len) {
+      val update: Int = primitiveArray(i).toInt
+      result = 37 * result + update
+      i += 1
+    }
+    result
+  }
+}
+
+final class GenericByteArrayData(val primitiveArray: Array[Byte])
+  extends GenericArrayData {
+  override def array(): Array[Any] = primitiveArray.toArray
+
+  override def copy(): ArrayData = new GenericByteArrayData(primitiveArray)
+
+  override def numElements(): Int = primitiveArray.length
+
+  override def isNullAt(ordinal: Int): Boolean = false
+  override def getByte(ordinal: Int): Byte = primitiveArray(ordinal)
+  override def toByteArray(): Array[Byte] = {
+    val array = new Array[Byte](numElements)
+    System.arraycopy(primitiveArray, 0, array, 0, numElements)
+    array
+  }
+  override def toString(): String = primitiveArray.mkString("[", ",", "]")
+
+  override def equals(o: Any): Boolean = {
+    if (!o.isInstanceOf[GenericByteArrayData]) {
+      return false
+    }
+
+    val other = o.asInstanceOf[GenericByteArrayData]
+    if (other eq null) {
+      return false
+    }
+
+    val len = numElements()
+    if (len != other.numElements()) {
+      return false
+    }
+
+    var i = 0
+    while (i < len) {
+      val o1 = primitiveArray(i)
+      val o2 = other.primitiveArray(i)
+      if (o1 != o2) {
+        return false
+      }
+      i += 1
+    }
+    true
+  }
+
+  override def hashCode: Int = {
+    var result: Int = 37
+    var i = 0
+    val len = numElements()
+    while (i < len) {
+      val update: Int = primitiveArray(i).toInt
+      result = 37 * result + update
+      i += 1
+    }
+    result
+  }
+}
+
+final class GenericBooleanArrayData(val primitiveArray: Array[Boolean])
```

Comment:
If possible, I would like to use ```Array[Boolean]```.

From the memory footprint view, it would be good to use [BitSet](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/collection/BitSet.scala). On the other hand, from the performance view, to keep this information by using ```Bitset``` instead of ```boolean[]``` may lead to data representation conversion.

If the generated code wants to access ```double[]``` (e.g. pass it to a lambda function for Dataset), to use ```Bitset``` would cause performance overhead.

What do you think?